domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt

```diff
@@ -172,6 +187,37 @@ internal class StateDeck internal constructor(
       .build()
   }
 
+  /**
+   * Returns an [ExplorationCheckpoint] which contains all the latest values of variables of the
+   * [StateDeck] that are used in light weight checkpointing.
+   */
+  internal fun createExplorationCheckpoint(
+    explorationVersion: Int,
+    explorationTitle: String,
+    timestamp: Long
+  ): ExplorationCheckpoint {
+    val explorationCheckpointBuilder = ExplorationCheckpoint.newBuilder()
+
+    previousStates.forEach { state ->
+      explorationCheckpointBuilder.addCompletedStatesInCheckpoint(
+        CompletedStateInCheckpoint.newBuilder()
+          .setCompletedState(state.completedState)
+          .setStateName(state.state.name)
+      )
+    }
+
+    return explorationCheckpointBuilder
+      .setPendingStateName(pendingTopState.name)
+      .setHintIndex(revealedHintIndex)
+      .setSolutionIsRevealed(solutionIsRevealed)
+      .addAllPendingUserAnswers(currentDialogInteractions)
+      .setStateIndex(stateIndex)
+      .setExplorationVersion(explorationVersion)
+      .setExplorationTitle(explorationTitle)
+      .setTimestampOfFirstCheckpoint(timestamp)
+      .build()
```

Comment:
```suggestion
    return ExplorationCheckpoint.newBuilder().apply {
      addAllCompletedStatesInCheckpoint(previousStates.map { state ->
        CompletedStateInCheckpoint.newBuilder().apply {
          completedState = state.completedState
          stateName = state.state.name
        }.build()
      })
      pendingStateName = pendingTopState.name
      hintIndex = revealedHintIndex
      this.solutionIsRevealed = solutionIsRevealed
      addAllPendingUserAnswers(currentDialogInteractions)
      this.stateIndex = stateIndex
      this.explorationVersion = explorationVersion
      this.explorationTitle = explorationTitle
      timestampOfFirstCheckpoint = timestamp
    }.build()
```

More Kotlin-idiomatic (per the ``map``), and conveniently in one expression to avoid saving the builder to a val (per the ``apply``).

Reply:
done

Reply:
The expression `this.stateIndex = stateIndex`  is being interpreted as `ExplorationCheckpointBuilder.stateIndex =  ExplorationCheckpointBuilder.stateIndex`

So reverting back to original.

Reply:
Latest changes now include this per discussion in the team meeting.